### PR TITLE
[DEVOPS-4156] Generate Statuscake config from Manifests

### PR DESCRIPTION
**Pull request by florent-babylon**

### CODE REVIEW

File: shipcat_cli/src/statuscake.rs [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+/// One Statuscake object
+#[derive(Serialize, Deserialize)]
+struct StatuscakeTest {
+    pub name: String,
+    pub WebsiteName: String,
+    pub WebsiteURL: String,
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub ContactGroup: Option<String>,
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub TestTags: Option<String>,
```
Comment (clux) [inline shipcat_cli/src/statuscake.rs]: snake_case in outputs please. we never use PascalCase anywhere afaik

File: shipcat_cli/src/statuscake.rs [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    //}
+
+    Ok(tests)
+}
+
+/// Generate Statuscake config from a filled in global config
+pub fn output(conf: &Config, region: &Region) -> Result<()> {
+    let res = generate_statuscake_output(&conf, &region)?;
+    let output = serde_yaml::to_string(&res)?;
+    let _ = io::stdout().write(format!("{}\n", output).as_bytes());
```
Comment (clux) [inline shipcat_cli/src/statuscake.rs]: you can just `println!("{}", output);` and remove the `io` import

File: shipcat_cli/src/statuscake.rs [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    // Generate list of APIs to feed to Statuscake
+    for svc in Manifest::available(&region.name)? {
+        debug!("Scanning service {:?}", svc);
+        let mf = Manifest::simple(&svc, &conf, region)?; // does not need secrets
+        debug!("Found service {} in region {}", mf.name, region.name);
+        if let Some(k) = mf.kong.clone() {
+           tests.push(StatuscakeTest::new(
+                   region,
+                   svc,
+                   k));
```
Comment (clux) [inline shipcat_cli/src/statuscake.rs]: k? does this compile?

File: shipcat_cli/src/statuscake.rs [hunk 4]
```diff
@@ -0,0 +1,80 @@
+use std::io::{self, Write};
+
+use super::{Manifest, Result, Region, Config};
+
+/// One Statuscake object
+#[derive(Serialize, Deserialize)]
```
[the remark this answers is not in the document]
Comment (florent-babylon) [reply shipcat_cli/src/statuscake.rs]: Indeed 👍

File: shipcat_cli/src/statuscake.rs [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    pub TestTags: Option<String>,
+}
+
+impl StatuscakeTest {
+    fn new(region: &Region, name: String, kong: Kong) -> Self {
+        let websiteName = format!("{} {} healthcheck", region.name, name);
+        // TODO HANDLE HOSTS
+        let websiteUrl = format!("{}/status/{}/health",
+                                 region.base_urls.get("external_services").unwrap(),
+                                 name);
```
[the remark this answers is not in the document]
Comment (florent-babylon) [reply shipcat_cli/src/statuscake.rs]: Yup, that was mostly to prototype :) Brought that up in the calling function, with handling of the error case

File: shipcat_cli/src/statuscake.rs [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+        // Generate tags, both regional and environment
+        let testTags = [
+            region.clone().name,
+            region.clone().environment
```
[the remark this answers is not in the document]
Comment (florent-babylon) [reply shipcat_cli/src/statuscake.rs]: Done! :)

File: shipcat_cli/src/statuscake.rs (hunk 2 above)
Comment (florent-babylon) [reply shipcat_cli/src/statuscake.rs]: 👌

File: shipcat_cli/src/statuscake.rs (hunk 1 above)
Comment (clux) [reply shipcat_cli/src/statuscake.rs]: `#[serde(rename_all = "PascalCase")]`

File: shipcat_cli/src/statuscake.rs [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+                    .collect::<Vec<&str>>()
+                    .first()
+                    .unwrap())
+        } else if let Some(uris) = kong.uris {
+            format!("{}/status/{}/health",
+                    external_svc,
+                    uris.trim_start_matches("/"))
+        } else {
+            // No host, no uri, what's going on?
+            format!("")
```
Comment (clux) [inline shipcat_cli/src/statuscake.rs]: `None` is better

File: shipcat_cli/src/statuscake.rs [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+        ].join(",");
+
+        // Process extra region-specific config
+        // Set the Contact group if available
+        let contact_group = if let Some(ref conf) = region.statuscake {
+            if let Some(ref region_tags) = conf.extra_tags {
+                test_tags = [
+                    test_tags,
+                    region_tags.to_string()
+                ].join(",");
```
Comment (clux) [inline shipcat_cli/src/statuscake.rs]: format!("{},{}", a, b)

File: shipcat_cli/src/statuscake.rs [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+        } else {
+            None
+        };
+
+        StatuscakeTest {
+            name: name.into(),
+            WebsiteName: website_name,
+            WebsiteURL: website_url,
+            ContactGroup: contact_group,
+            TestTags: Some(test_tags),
```
Comment (clux) [inline shipcat_cli/src/statuscake.rs]: change struct name to snake case, use rename all, then these can use short hand initialization:
StatusCakeTest { website_name, website_url, contact_group, name: name.into(), ...}

File: shipcat_cli/src/statuscake.rs (hunk 8 above)
Comment (florent-babylon) [reply shipcat_cli/src/statuscake.rs]: 👍

File: shipcat_cli/src/statuscake.rs (hunk 9 above)
Comment (florent-babylon) [reply shipcat_cli/src/statuscake.rs]: 👍

File: shipcat_cli/src/statuscake.rs (hunk 1 above)
Comment (florent-babylon) [reply shipcat_cli/src/statuscake.rs]: 👍

File: shipcat_cli/src/statuscake.rs (hunk 7 above)
Comment (florent-babylon) [reply shipcat_cli/src/statuscake.rs]: Indeed, making this into an `Option` all the way.

File: shipcat_cli/src/statuscake.rs [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+struct StatuscakeTest {
+    #[serde(rename = "name")]
+    pub name: String,
+    pub website_name: String,
+    #[serde(default, rename = "WebsiteURL")]
+    pub website_url: Option<String>,
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub contact_group: Option<String>,
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub test_tags: Option<String>,
```
Comment (clux) [inline shipcat_cli/src/statuscake.rs]: this doesn't need to be in an option, you're always returning it

File: shipcat_cli/src/statuscake.rs [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+#[serde(rename_all = "PascalCase")]
+struct StatuscakeTest {
+    #[serde(rename = "name")]
+    pub name: String,
+    pub website_name: String,
+    #[serde(default, rename = "WebsiteURL")]
+    pub website_url: Option<String>,
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub contact_group: Option<String>,
+    #[serde(default, skip_serializing_if = "Option::is_none")]
```
Comment (clux) [inline shipcat_cli/src/statuscake.rs]: `default` serde annotations are not needed in this struct, you are never Deserializing

File: shipcat_definitions/src/region.rs [hunk 12]
```diff
@@ -107,6 +107,18 @@ pub struct KongConfig {
...
+/// StatusCake configuration for a region
+#[derive(Serialize, Deserialize, Clone, Default)]
+#[serde(deny_unknown_fields)]
+pub struct StatuscakeConfig {
+    /// Contact Group that will be used if tests go down
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub contact_group: Option<String>,
+    /// Extra tags to add to all tests in this region
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub extra_tags: Option<String>,
```
Comment (clux) [inline shipcat_definitions/src/region.rs]: much optional, i assume it's necessary

File: shipcat_definitions/src/region.rs [hunk 13]
```diff
@@ -235,6 +247,9 @@ pub struct Region {
     /// Kong configuration for the region
     #[serde(default)]
     pub kong: KongConfig,
+    /// Statuscake configuration for the region
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub statuscake: Option<StatuscakeConfig>,
```
Comment (clux) [inline shipcat_definitions/src/region.rs]: :+1:

File: shipcat_cli/src/statuscake.rs (hunk 10 above)
Comment (florent-babylon) [reply shipcat_cli/src/statuscake.rs]: Yeah, that's true actually :) Making it non-optional

File: shipcat_definitions/src/region.rs (hunk 12 above)
Comment (florent-babylon) [reply shipcat_definitions/src/region.rs]: No indeed - making it necessary 👍

File: shipcat_cli/src/statuscake.rs (hunk 11 above)
Comment (florent-babylon) [reply shipcat_cli/src/statuscake.rs]: 👍

**Pull request merged**